core/src/Cuda/Kokkos_CudaSpace.cpp

```diff
@@ -161,52 +174,36 @@ void *CudaSpace::allocate(const char *arg_label, const size_t arg_alloc_size,
 }
 
 namespace {
-void *impl_allocate_common(const Cuda &exec_space, const char *arg_label,
-                           const size_t arg_alloc_size,
+void *impl_allocate_common(const int device_id,
+                           [[maybe_unused]] const cudaStream_t stream,
+                           const char *arg_label, const size_t arg_alloc_size,
                            const size_t arg_logical_size,
                            const Kokkos::Tools::SpaceHandle arg_handle,
-                           bool exec_space_provided) {
+                           [[maybe_unused]] bool stream_sync_only) {
   void *ptr = nullptr;
+  KOKKOS_IMPL_CUDA_SAFE_CALL(cudaSetDevice(device_id));
 
+  cudaError_t error_code = cudaSuccess;
 #ifndef CUDART_VERSION
 #error CUDART_VERSION undefined!
 #elif (defined(KOKKOS_ENABLE_IMPL_CUDA_MALLOC_ASYNC) && CUDART_VERSION >= 11020)
-  cudaError_t error_code;
   if (arg_alloc_size >= memory_threshold_g) {
-    if (exec_space_provided) {
-      error_code =
-          exec_space.impl_internal_space_instance()->cuda_malloc_async_wrapper(
-              &ptr, arg_alloc_size);
-      exec_space.fence("Kokkos::Cuda: backend fence after async malloc");
+    error_code = cudaMallocAsync(&ptr, arg_alloc_size, stream);
+
+    if (stream_sync_only) {
+      KOKKOS_IMPL_CUDA_SAFE_CALL(cudaStreamSynchronize(stream));
```

Comment:
There is an `Impl::cuda_stream_synchronize`

Reply:
Also I don't fully understand the error code handling in that function.

Reply:
> There is an Impl::cuda_stream_synchronize

Yes, but that would set the device to the default one which would be wrong here.

Reply:
> Also I don't fully understand the error code handling in that function.

Which function? This one? I tried to preserve behavior.

Reply:
The only way this function could fail is if an invalid stream was provided. Maybe it's better to guard with
```
if (error_code == cudaSuccess) {
if (stream_sync_only) 
[...]
else
[...]
}
```
instead.

Reply:
00f0a73747d89c4402678f5c3360c7195d245e84